src/mlmc/mc_level.py

```diff
@@ -1,190 +1,431 @@
+        """
+        Add samples pair to rest of samples
+        :param id: sample id
+        :param sample_pair: Fine and coarse result
+        :return: None
+        """
+        fine, coarse = sample_pair
+        # Samples are not finite
+        if not np.isfinite(fine) or not np.isfinite(coarse):
+            self.nan_samples.append(id)
+            return
+        # Enlarge matrix of samples
+        if self._n_valid_samples == self._sample_values.shape[0]:
+            self.enlarge_samples(2*self._n_valid_samples)
 
-    def get_moments(self):
+        # Add fine and coarse sample
+        self._sample_values[self._n_valid_samples, :] = (fine, coarse)
+        self._n_valid_samples += 1
+
+    def enlarge_samples(self, size):
         """
-        Get moments from results of simulations on this level
-        :return: array, moments
+        Enlarge matrix of samples
+        :param size: New sample matrix size
+        :return: None
         """
-        self.moments = self.level_moments()
-        if len(self.moments_estimate) == 0:
-            self.moments_estimate = self._moments
+        # Enlarge sample matrix
+        new_values = np.empty((size, 2))
+        new_values[:self._n_valid_samples] = self._sample_values[:self._n_valid_samples]
+        self._sample_values = new_values
 
+    @property
+    def n_total_samples(self):
+        """
+        Number of all level samples
+        :return: int
+        """
+        return len(self.running_simulations) + len(self.finished_simulations)
+
+    @property
+    def n_samples(self):
+        # Number of samples used for estimates.
+        if self.sample_indices is None:
+            return self._n_valid_samples
+        else:
+            return len(self.sample_indices)
+
+    def _get_sample_tag(self, char):
+        """
+        Create sample tag
+        :param char: 'C' or 'F' depending on the type of simulation
+        :return: str
+        """
+        return "L{:02d}_{}_S{:07d}".format(self._logger.level_idx, char, self.n_total_samples)
+
+    @property
     def n_ops_estimate(self):
         """
-        :return: fine simulation n
+        :return: number of fine sim operations
+        """
+        return self._n_ops_estimate
+
+    @n_ops_estimate.setter
+    def n_ops_estimate(self, n_ops):
+        """
+        Set n ops estimate
+        :param n_ops: number of operations
+        :return: None
+        """
+        if self._logger.n_ops_estimate is None or self._logger.n_ops_estimate > 0:
+            self._n_ops_estimate = n_ops
+            self._logger.n_ops_estimate = n_ops
+        
+    def set_coarse_sim(self):
         """
-        return self.fine_simulation.n_ops_estimate()
+        Set coarse sim to fine simulation
+        :return: None
+        """
+        if not self.fine_simulation.coarse_sim_set:
+            self.fine_simulation.set_coarse_sim(self.coarse_simulation)
+            self.n_ops_estimate = self.fine_simulation.n_ops_estimate()

```

Comment:
It is because of generating samples in reverse order (from the finest), but simulations have to be created in order from first level to last, so that is the reason why I move this code into own method which is called separately in mlmc.refill_samples()